File: 2559-maximum-number-of-non-overlapping-palindrome-substrings/maximum-number-of-non-overlapping-palindrome-substrings.py
class Solution:
    def maxPalindromes(self, s: str, k: int) -> int:
        n = len(s)
        count = 0
        i = 0

        def is_palindrome(l: int, r: int) -> bool:
            if r >= n:
                return False
            while l < r:
                if s[l] != s[r]:
                    return False
                l += 1
                r -= 1
            return True

        while i < n:
            found = False
            for j in range(i, n):
                if is_palindrome(j, j + k - 1):
                    count += 1
                    i = j + k
                    found = True
                    break
                if is_palindrome(j, j + k):
                    count += 1
                    i = j + k + 1
                    found = True
                    break
            if not found:
                break

        return count

File: 2559-maximum-number-of-non-overlapping-palindrome-substrings/maximum-number-of-non-overlapping-palindrome-substrings.java
class Solution {
    public int maxPalindromes(String s, int k) {
        int n = s.length();
        int count = 0;
        int i = 0;

        while (i < n) {
            boolean found = false;
            for (int j = i; j < n; j++) {
                if (isPalindrome(s, j, j + k - 1)) {
                    count++;
                    i = j + k;
                    found = true;
                    break;
                }
                if (isPalindrome(s, j, j + k)) {
                    count++;
                    i = j + k + 1;
                    found = true;
                    break;
                }
            }
            if (!found) break;
        }

        return count;
    }

    private boolean isPalindrome(String s, int l, int r) {
        if (r >= s.length()) return false;
        while (l < r) {
            if (s.charAt(l) != s.charAt(r)) return false;
            l++;
            r--;
        }
        return true;
    }
}